aptos-move/aptos-vm/src/move_vm_ext/respawned_session.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    aptos_vm_impl::AptosVMImpl,
+    data_cache::{AsMoveResolver, StorageAdapter},
+    move_vm_ext::{SessionExt, SessionId},
+};
+use anyhow::{bail, Result};
+use aptos_aggregator::transaction::ChangeSetExt;
+use aptos_gas::ChangeSetConfigs;
+use aptos_state_view::{StateView, StateViewId, TStateView};
+use aptos_types::{
+    state_store::{
+        state_key::StateKey, state_storage_usage::StateStorageUsage, state_value::StateValue,
+    },
+    write_set::{TransactionWrite, WriteSet},
+};
+use move_core_types::vm_status::{err_msg, StatusCode, VMStatus};
+
+/// We finish the session after the user transaction is done running to get the change set and
+/// charge gas and storage fee based on it before running storage refunds and the transaction
+/// epilogue. The latter needs to see the state view as if the change set is applied on top of
+/// the base state view, and this struct implements that.
+#[ouroboros::self_referencing]
+pub struct RespawnedSession<'r, 'l> {
+    state_view: ChangeSetStateView<'r>,
+    #[borrows(state_view)]
+    #[covariant]
+    resolver: StorageAdapter<'this, ChangeSetStateView<'r>>,
+    #[borrows(resolver)]
+    #[not_covariant]
+    session: Option<SessionExt<'this, 'l>>,
+}
+
+impl<'r, 'l> RespawnedSession<'r, 'l> {
+    pub fn spawn(
+        vm: &'l AptosVMImpl,
+        session_id: SessionId,
+        base_state_view: &'r dyn StateView,
+        previous_session_change_set: ChangeSetExt,
+    ) -> Result<Self, VMStatus> {
+        let state_view = ChangeSetStateView::new(base_state_view, previous_session_change_set)?;
+
+        Ok(RespawnedSessionBuilder {
+            state_view,
+            resolver_builder: |state_view| state_view.as_move_resolver(),
+            session_builder: |resolver| Some(vm.new_session(resolver, session_id)),
+        }
+        .build())
+    }
+
+    pub fn execute<R>(&mut self, fun: impl FnOnce(&mut SessionExt) -> R) -> R {
+        self.with_session_mut(|session| fun(session.as_mut().unwrap()))
+    }
```

Comment:
Would it be cleaner to just implement Deref?

Reply:
You mean DerefMut? And I can't.. the struct. Ouroboros doesn't generate `.borrow_FIELD_mut()` https://docs.rs/ouroboros/latest/ouroboros/attr.self_referencing.html (search for "There is no borrow_FIELD_mut") (I'm not good enough to figure out exactly why)